beacon_node/beacon_chain/src/eth1_chain.rs

```diff
@@ -1,27 +1,90 @@
-use crate::BeaconChainTypes;
+use eth1::{Config as Eth1Config, Eth1Block, Service as HttpService};
 use eth2_hashing::hash;
+use exit_future::Exit;
+use futures::Future;
+use integer_sqrt::IntegerSquareRoot;
+use rand::prelude::*;
+use slog::{crit, Logger};
+use std::collections::HashMap;
+use std::iter::DoubleEndedIterator;
+use std::iter::FromIterator;
 use std::marker::PhantomData;
-use types::{BeaconState, Deposit, Eth1Data, EthSpec, Hash256};
+use std::sync::Arc;
+use store::{Error as StoreError, Store};
+use types::{
+    BeaconState, BeaconStateError, ChainSpec, Deposit, Eth1Data, EthSpec, Hash256, Slot, Unsigned,
+};
 
-type Result<T> = std::result::Result<T, Error>;
+const DEPOSIT_TREE_DEPTH: usize = 32;
+
+type BlockNumber = u64;
+type Eth1DataBlockNumber = HashMap<Eth1Data, BlockNumber>;
+type Eth1DataVoteCount = HashMap<(Eth1Data, BlockNumber), u64>;
+
+#[derive(Debug, PartialEq)]
+pub enum Error {
+    /// Unable to return an Eth1Data for the given epoch.
+    EpochUnavailable,
+    /// An error from the backend service (e.g., the web3 data fetcher).
+    BackendError(String),
+    /// The deposit index of the state is higher than the deposit contract. This is a critical
+    /// consensus error.
+    DepositIndexTooHigh,
+    /// The current state was unable to return the root for the state at the start of the eth1
+    /// voting period.
+    UnableToGetPreviousStateRoot(BeaconStateError),
+    /// The state required to find the previous eth1 block was not found in the store.
+    PreviousStateNotInDB,
+    /// There was an error accessing an object in the database.
+    StoreError(StoreError),
+    /// The eth1 head block at the start of the eth1 voting period is unknown.
+    ///
+    /// The eth1 caches are likely stale.
+    UnknownVotingPeriodHead,
+    /// The block that was previously voted into the state is unknown.
+    ///
+    /// The eth1 caches are stale, or a junk value was voted into the chain.
+    UnknownPreviousEth1BlockHash,
+}
 
 /// Holds an `Eth1ChainBackend` and serves requests from the `BeaconChain`.
-pub struct Eth1Chain<T: BeaconChainTypes> {
-    backend: T::Eth1Chain,
+pub struct Eth1Chain<T, E>
+where
+    T: Eth1ChainBackend<E>,
+    E: EthSpec,
+{
+    backend: T,
+    /// When `true`, the backend will be ignored and dummy data from the 2019 Canada interop method
+    /// will be used instead.
+    pub use_dummy_backend: bool,
```

Comment:
I totally agree with you conceptually and originally I had it the way you suggested. _But_, it turns out that including it at the type level requires significantly more code on the run-time side and this hack seemed like a fair trade off to me.

Most of my reasoning is that I don't really see a reason for `DummyEth1Backend` to exist; voting on deterministic not-real-data in order to keep the deposit count stable is equivalent to just voting completely random junk. That being said, I'm not quite sure I'm ready to raise this up the "client interop" flagpole yet, so I thought we'd leave it in here for now.

We can chat about this in person today, but my feeling is to leave this as-is for now instead of (a) spending the time on a refactor and (b) introducing more code that isn't run during tests but is used at run-time.